packages/beacon-node/src/chain/validation/attestation.ts

```diff
@@ -568,41 +571,111 @@ export function verifyHeadBlockAndTargetRoot(
+  } else {
+    chain.metrics?.gossipAttestation.shufflingRegenMiss.inc({caller: regenCaller});
+    return null;
+  }
+}
+
+/**
+ * Get dependent root of a shuffling given attestation epoch and head block.
+ */
+export function getShufflingDependentRoot(
+  chain: IBeaconChain,
+  attEpoch: Epoch,
+  blockEpoch: Epoch,
+  attHeadBlock: ProtoBlock
+): RootHex {
+  let shufflingDependentRoot: RootHex;
+  if (blockEpoch === attEpoch) {
+    // current shuffling, this is equivalent to `headState.currentShuffling`
+    // given blockEpoch = attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     ------------------------^
+    // shufflingDependentRoot ------^
+    shufflingDependentRoot = chain.forkChoice.getDependentRoot(attHeadBlock, EpochDifference.previous);
+  } else if (blockEpoch === attEpoch - 1) {
+    // next shuffling, this is equivalent to `headState.nextShuffling`
+    // given blockEpoch = n-1, attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     -------------------^
+    // shufflingDependentRoot ------^
+    shufflingDependentRoot = chain.forkChoice.getDependentRoot(attHeadBlock, EpochDifference.current);
+  } else if (blockEpoch < attEpoch - 1) {
+    // this never happens with default chain option of maxSkipSlots = 32, however we still need to handle it
+    // check the verifyHeadBlockAndTargetRoot() function above
+    // given blockEpoch = n-2, attEpoch = n
+    //        epoch:       (n-2)   (n-1)     n     (n+1)
+    //               |-------|-------|-------|-------|
+    // attHeadBlock     -----------^
+    // shufflingDependentRoot -----^
+    shufflingDependentRoot = attHeadBlock.blockRoot;
+    // use lodestar_gossip_attestation_head_slot_to_attestation_slot metric to track this case
+  } else {
+    // blockEpoch > attEpoch
+    // should not happen, handled in verifyAttestationTargetRoot
+    throw Error(`attestation epoch ${attEpoch} is before head block epoch ${blockEpoch}`);
+  }
+
+  return shufflingDependentRoot;
+}
+
+/**
+ * Different version of getAttestationDataSigningRoot in state-transition which doesn't require a state.
+ */
+export function getAttestationDataSigningRoot(config: BeaconConfig, data: phase0.AttestationData): Uint8Array {
+  const slot = computeStartSlotAtEpoch(data.target.epoch);
+  // previously, we call `domain = config.getDomain(state.slot, DOMAIN_BEACON_ATTESTER, slot)`
+  // at fork boundary, it's required to dial to target epoch https://github.com/ChainSafe/lodestar/blob/v1.11.3/packages/beacon-node/src/chain/validation/attestation.ts#L573
+  // instead of that, just use the fork at slot in the attestation data
+  const fork = config.getForkName(slot);
+  const domain = config.getDomainAtFork(fork, DOMAIN_BEACON_ATTESTER);
+  return computeSigningRoot(ssz.phase0.AttestationData, data, domain);
 }
```

Comment:
previously I used `config.getDomain()` api, and used message slot at the position of state slot which was a hacky way. Now I changed to `config.getDomainAtFork()` which make it more straightforward

I also double checked lighthouse's implementation https://github.com/sigp/lighthouse/blob/v4.5.0/beacon_node/beacon_chain/src/attestation_verification.rs#L1183

need to review this logic again @wemeetagain @dapplion , let me know if I need to add more comments

Reply:
Logic looks correct! My previous comment did not realize that the complexity on picking the right fork is on the state transition not gossip